librad/src/id/entity.rs

```diff
@@ -0,0 +1,614 @@
+        key: &PublicKey,
+        by: &Signatory,
+        signature: &Signature,
+        resolver: &impl Resolver<User>,
+    ) -> Result<(), Error> {
+        self.check_key(key, by, resolver).await?;
+        if signature.verify(&self.canonical_data()?, key) {
+            Ok(())
+        } else {
+            Err(Error::SignatureVerificationFailed)
+        }
+    }
+
+    /// Check the validity of this entty (only this revision is checked)
+    ///
+    /// This checks that:
+    /// - every owned key and certifier has a corresponding signature
+    /// - only owned keys and certifiers have signed the entity
+    pub async fn check_validity(&self, resolver: &impl Resolver<User>) -> Result<(), Error> {
+        let mut keys = HashSet::<PublicKey>::from_iter(self.keys().iter().cloned());
+        let mut users = HashSet::<RadicleUri>::from_iter(self.certifiers().iter().cloned());
+
+        for (k, s) in self.signatures() {
+            self.check_signature(k, &s.by, &s.sig, resolver).await?;
+            match &s.by {
+                Signatory::OwnedKey => {
+                    keys.remove(k);
+                },
+                Signatory::User(user) => {
+                    users.remove(&user);
+                },
+            }
+        }
+        if keys.is_empty() && users.is_empty() {
+            Ok(())
+        } else {
+            Err(Error::SignatureMissing)
+        }
+    }
+
+    /// Convenience method to check if an entity is valid
+    pub async fn is_valid(&self, resolver: &impl Resolver<User>) -> bool {
+        self.check_validity(resolver).await.is_ok()
+    }
+
+    /// Given an entity and its previous revision check that the update is
+    /// valid:
+    ///
+    /// - the revision has been incremented
+    /// - the parent hash is correct
+    /// - the TUF quorum rules have been observed
+    ///
+    /// FIXME: only allow exact `+1`increments so that the revision history has
+    /// no holes
+    /// FIXME: probably we should merge owned keys and certifiers when checking
+    /// the quorum rules (now we are handling them separately)
+    pub fn check_update(&self, previous: &Self) -> Result<(), UpdateVerificationError> {
+        if self.revision() <= previous.revision() {
+            return Err(UpdateVerificationError::NonMonotonicRevision);
+        }
+
+        match &self.parent_hash {
+            Some(parent_hash) => {
+                if &previous.hash != parent_hash {
+                    return Err(UpdateVerificationError::WrongParentHash);
+                }
+            },
+            None => {
+                return Err(UpdateVerificationError::WrongParentHash);
+            },
+        }
+
+        let retained_keys = self.keys().iter().filter(|k| previous.has_key(k)).count();
+        let total_keys = self.keys_count();
+        let added_keys = total_keys - retained_keys;
+        let removed_keys = previous.keys_count() - retained_keys;
+        let quorum_keys = total_keys / 2;
```

Comment:
I wonder if we want the threshold to be configurable?

Reply:
I wouldn't do this now -- TUF mandates to specify the threshold, but I left this out when writing the spec intentionally because it is something that will need to be carefully communicated in the UI: it is probably hard enough to understand why this thing is asking you for your signature, but then you also have to understand what implications it might have to change this one number.